File: web/tests/reference/queues/queue-using-stacks.py
class MyQueue:
    def __init__(self):
        self.inbox = []
        self.outbox = []

    def _pour(self):
        if not self.outbox:
            while self.inbox:
                self.outbox.append(self.inbox.pop())

    def push(self, x):
        self.inbox.append(x)

    def pop(self):
        self._pour()
        return self.outbox.pop()

    def peek(self):
        self._pour()
        return self.outbox[-1]

    def empty(self):
        return not self.inbox and not self.outbox

File: web/tests/reference/queues/queue-using-stacks.js
class MyQueue {
  constructor() {
    this.inbox = [];
    this.outbox = [];
  }

  pour() {
    if (this.outbox.length === 0) {
      while (this.inbox.length > 0) this.outbox.push(this.inbox.pop());
    }
  }

  push(x) {
    this.inbox.push(x);
  }

  pop() {
    this.pour();
    return this.outbox.pop();
  }

  peek() {
    this.pour();
    return this.outbox[this.outbox.length - 1];
  }

  empty() {
    return this.inbox.length === 0 && this.outbox.length === 0;
  }
}
